### ISSUE TITLE
result of racket-send-last-sexp in repl buffer is confusing

### ISSUE THREAD

**Issue by cosmicexplorer** [greghendershott/racket-mode#171]
Evaluating a form before point with `racket-send-last-sexp` sends the form to the `*Racket REPL*` buffer and evaluates it, and leaves it in the repl buffer as if it was an expression the user entered. This leads to confusing behavior, as below:
`test.rkt`:
```
(print "hey")
(closure 1 2)
```
`*Racket REPL*`:
```
hw5.rkt﻿> "hey"
hw5.rkt﻿> (+ 1 2)
3
hw5.rkt﻿> (closure 1 2)
hw5.rkt﻿> 
```

Can you tell what happened here? In this case, the user evaluated the form `(print "hey")` in `test.rkt`, then switched to the repl buffer, then evaluated `(+ 1 2)` there, then evaluated `(closure 1 2)` in `test.rkt` (where `closure` is some random struct, sorry for the confusing name). This sequence of events is not immediately obvious from looking at the repl buffer, and normal keybindings like `M-p` or `M-n` don't traverse the history of `racket-send-last-sexp` results. So the current behavior leads to a confusing, inconsistent log of evaluation order in the repl buffer.

Is there a reason it's written this way? Other lispy interaction modes like [cider](https://github.com/clojure-emacs/cider) just display the result of `C-x C-e` in the minibuffer just like native emacs lisp expressions; if they log the results at all, they log them as "results;" i.e. not at the same line as the prompt. If the above were to be translated to cider's style, it would likely look like the below, with minibuffer display as well when the user evaluates each line:

```
"hey"
hw5.rkt﻿> (+ 1 2)
3
(closure 1 2)
hw5.rkt﻿> 
```

In this case, it's obvious that `"hey"` is the result of some evaluation run elsewhere, *not* in the repl buffer, as is the `(closure 1 2)` message. In addition, because of the minibuffer echoing as the expression is evaluated, it's not required that the user even have the repl buffer open for quick and dirty evaluations like `C-x C-e` is usually used for.

So is there a reason the current version works the way it does? If not I can try to whip up a pull request to enforce what I guess I'm calling cider-like behavior.

**Comment by greghendershott** [comment]
It's a good point.

This all got started as some Emacs Lisp I wrote to make `scheme-mode` and `inferior-scheme-mode` work in ways that seemed less weird, to me, coming from DrRacket and with no prior lisp experience. Eventually it evolved into its own distinct `racket-mode` and `racket-repl-mode`.

What you're describing is another bit of weirdness, just one that never bothered me enough. :)  I agree it could be better.

IIUC this is basically a consequence of using `comint-send-region` and `comint-send-string`, which doesn't echo things in the REPL. That's fine except for not echoing the newline. If that happened, the output would get its own, fresh line. Maybe it's that simple.

So I'll look at changing this.

---

p.s. I'm not a big fan of <kbd>C-x C-e</kbd> using the minibuffer. It's too fleeting for my taste. Instead I like having the result "logged" in the REPL buffer, where it's readily available to refer to later, copy and modify, etc.  I guess there are arguments to make either way, maybe it's a matter of what we're used to. :)

**Comment by cosmicexplorer** [comment]
I completely agree that <kbd>C-x C-e</kbd> should log everything in the REPL, for the same reasons you mentioned; I just don't like that it looks like an input instead of an output (I also just like the minibuffer, too). A solution which would satisfy both of us, I think, would be to copy the sexp <kbd>C-x C-e</kbd> is currently evaluating over to the REPL buffer, then run `comint-send-input`, which would make the expression available in input history. An alternative similar to the method I described in the OP would be to have some logic which moves the repl's point to before the current prompt, then runs `comint-send-region` so it appears as an "output" instead of an "input" (although that seems harder and for little benefit).

Thanks for taking this suggestion so seriously. If this is something you could see happening I could whip up maybe some code and a `defcustom` so it doesn't affect the default behavior, but can be switched on.

**Comment by greghendershott** [comment]
> A solution which would satisfy both of us, I think, would be to copy the sexp C-x C-e is currently evaluating over to the REPL buffer, then run comint-send-input, which would make the expression available in input history.

One drawback to copying and showing them in the REPL, would be that <kbd>C-x C-e</kbd> -- and especially <kbd>M-C-x</kbd> -- may be used with very large sexps. It could be distracting for these to appear in the REPL, especially if they're big enough to cause the window to scroll far.

> An alternative similar to the method I described in the OP would be to have some logic which moves the repl's point to before the current prompt, then runs comint-send-region so it appears as an "output" instead of an "input" (although that seems harder and for little benefit).

Although not exactly what you're describing here, I'm testing a `comint-preoutput-filter-functions` hook that checks if point is on the same line as a prompt, and if so, inserts a `\n` before the output.

[ICYI: It's better if this is only used for the few `send-xxx-to-repl` commands that we're talking about -- not generally. Otherwise it causes extraneous newlines and prompts when Racket is sending its own hidden/internal commands to the Racket REPL process.]

So far this is working well but I'd like to test some more before committing and pushing.

**Comment by greghendershott** [comment]
Actually the preoutput-filter hook isn't needed. It's sufficient to insert a `?\n` at the `process-mark` -- provided the marker is updated to follow the `?\n`. That was the missing piece that led me to muck with the preoutput-filter.

**Comment by cosmicexplorer** [comment]
Awesome! Thanks for making this change so quickly.

**Comment by greghendershott** [comment]
You're welcome. A year or two ago, this annoyed me, but I had much bigger fish to fry. And then I got desensitized.

Also I just don't do much `racket-send-last-sexp`. And especially not `racket-send-definition`; coming from DrRacket and from non-Lisps, my first instinct to update the original source file and re-run it. However it does come in handy. Especially now that racket-mode lets you properly enter Racket submodules, for example, this is handy to eval individual expressions in a `test` submodule. And so on.

TL;DR thanks for the reminder this needed to be de-weirded.
